### PR TITLE
Support specifying a similarity threshold

### PR DESCRIPTION
**Pull request by brylie**
Related to #21

Add support for specifying a similarity threshold in vector search methods.

* **VectorIndex Class**: Add `similarity_threshold` parameter to `query`, `find_similar`, and `search` methods in `src/wagtail_vector_index/storage/base.py`.
* **PgvectorIndexMixin Class**: Update `get_similar_documents` and `aget_similar_documents` methods to accept `similarity_threshold` parameter and filter results based on it in `src/wagtail_vector_index/storage/pgvector/provider.py`.
* **QdrantIndexMixin Class**: Update `get_similar_documents` method to accept `similarity_threshold` parameter and filter results based on it in `src/wagtail_vector_index/storage/qdrant/provider.py`.
* **WeaviateIndexMixin Class**: Update `get_similar_documents` method to accept `similarity_threshold` parameter and filter results based on it in `src/wagtail_vector_index/storage/weaviate/provider.py`.
* **NumpyIndexMixin Class**: Update `get_similar_documents` method to accept `similarity_threshold` parameter and filter results based on it in `src/wagtail_vector_index/storage/numpy/provider.py`.
* **Documentation**: Update `docs/vector-indexes.md` to include information on the new `similarity_threshold` parameter and provide examples of its usage.
* **Tests**: Add tests for the new `similarity_threshold` parameter in `query`, `find_similar`, and `search` methods in `tests/test_index.py`.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/wagtail/wagtail-vector-index/issues/21?shareId=c122e9f4-e046-4a9d-ad8c-9792d2c6cf9a).

### CODE REVIEW

**Comment by brylie** [comment]
I've marked this as a draft to indicate that it is work-in-progress, mainly due to failing tests. Early feedback is more than welcome :-)

**Comment by tomusher** [comment]
This is looking great thanks @brylie! I'll do a more thorough review this week and we get this in.

File: tests/test_index.py [hunk 1]
```diff
@@ -134,3 +204,80 @@ def get_similar_documents(query_embedding, limit=0):
...
+
+    def gen_pages(cls, *args, **kwargs):
+        yield from pages
+
+    mocker.patch(
+        "wagtail_vector_index.storage.models.EmbeddableFieldsDocumentConverter.bulk_from_documents",
+        side_effect=gen_pages,
+    )
+
+    case = unittest.TestCase()
```
Comment (tomusher) [inline tests/test_index.py:240]: Could we use [pytest-style asserts](https://docs.pytest.org/en/7.1.x/how-to/assert.html#asserting-with-the-assert-statement) here?

File: tests/test_index.py [hunk 2]
```diff
@@ -4,15 +4,85 @@
...
+        ExamplePageFactory(title="Somewhat similar to test"),
+        ExamplePageFactory(title="Not similar at all"),
+    ]
+
+
+@pytest.fixture
+def document_generator(test_pages):
+    def gen_documents(cls, *args, **kwargs):
+        for page in test_pages:
+            if "Very similar" in page.title:
```
Comment (tomusher) [inline tests/test_index.py:50]: **Suggestion** Is it worth moving this to use the same mocked `embed` function so we don't have to repeat the defined vectors?

**Review by tomusher — COMMENTED** [review]
Looks great to me, appreciate all the effort that went in to documentation. Left a couple of minor comments.

File: tests/test_index.py (hunk 2 above)
Comment (brylie) [reply tests/test_index.py:50]: Good suggestion. Done

File: tests/test_index.py (hunk 1 above)
Comment (brylie) [reply tests/test_index.py]: Done

**Comment by tomusher** [comment]
Thanks @brylie!

**Pull request merged**